R/groupingsets.R

```diff
@@ -66,6 +74,44 @@ groupingsets.data.table = function(x, j, by, sets, .SDcols, id = FALSE, jj, ...)
     stopf("Character vectors in 'sets' list must not have duplicated column names within a single grouping set.")
   if (length(sets) > 1L && (idx<-anyDuplicated(lapply(sets, sort))))
     warningf("'sets' contains a duplicate (i.e., equivalent up to sorting) element at index %d; as such, there will be duplicate rows in the output -- note that grouping by A,B and B,A will produce the same aggregations. Use `sets=unique(lapply(sets, sort))` to eliminate duplicates.", idx)
+  if (is.list(label)) {
+    other.allowed.names = c("character", "integer", "numeric", "factor", "Date", "IDate")
+    allowed.label.list.names = c(by, vapply_1c(.shallow(x, by), function(u) class(u)[1]),
+                                 other.allowed.names)
+    label.names = names(label)
+    if (!all(label.names %in% allowed.label.list.names))
+      stopf("When argument 'label' is a list, all element names must be (1) in 'by', or (2) the first element of the class in the data.table 'x' of a variable in 'by', or (3) one of %s. Element names not satisfying this condition: %s",
+            brackify(other.allowed.names), brackify(setdiff(label.names, allowed.label.list.names)))
+    label.classes = lapply(label, class)
+    label.names.in.by = intersect(label.names, by)
+    label.names.not.in.by = setdiff(label.names, label.names.in.by)
+    label.names.in.by.classes = label.classes[label.names.in.by]
+    x.label.names.in.by.classes = lapply(.shallow(x, label.names.in.by), class)
+    label.names.in.by.classes.match = vapply_1b(label.names.in.by,
+                                                function(u) identical(label.names.in.by.classes[[u]],
+                                                                      x.label.names.in.by.classes[[u]]))
+    label.names.not.in.by.classes1 = vapply_1c(label.classes[label.names.not.in.by], function(u) u[1])
+    label.names.not.in.by.classes1.match = (label.names.not.in.by == label.names.not.in.by.classes1)
+    if (!all(label.names.in.by.classes.match)) {
+      label.names.in.by.classes.mismatch.info =
+        paste0(label.names.in.by[!label.names.in.by.classes.match],
```

Comment:
See #6503, #6482 -- this approach to building the error message makes it hard to translate.

Reply:
(resolving with be0377680d4dd266ef552ea1c43d2710277aa067)